### DISCUSSION TITLE
How to read 24 bits frames with the I2S driver

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#9252] [Libraries & Drivers] [answered]
The [docs](https://docs.micropython.org/en/latest/library/machine.I2S.html#machine-i2s) of I2S state that the allowed sample sizes are 16 bits or 32 bits.

**How can we handle dataframes of 24 bits without too much manipulation?**

In the ["old" I2S-examples repo](https://github.com/miketeachman/micropython-esp32-i2s-examples) of @miketeachman (prior to the merge in the official Micropython codebase) you could use for example ```dataformat=I2S.B24```.

**Comment by miketeachman** [comment, marked as answer]
Can you describe what I2S device you are using that requires 24 bit audio samples?

**Reply by saraverbeecke** [reply]
I have an INMP441 microphone like the one you tested in your "old" I2S repo. I would like to use the I2S of Micropython version 1.19.1. How could we read out 24 bits format with that version?

**Reply by miketeachman** [reply]
Setting `bits=32` in the `machine.I2S` constructor will allow the `read()` method to read all 24 bits of audio data from the INMP441 microphone.  The 3 most significant bytes of the 32 bit value will contain the 24 bits of sample data, while the lower byte will contain 0s.  

One important consideration to make sure the lower byte contains 0s:  The datasheet for the INMP441 indicates:

> The SD trace should have a pull-down resistor to discharge the line during the time that all microphones on the bus have tri-stated
> their outputs. A 100 kΩ resistor is sufficient for this, as shown in Figure 7.

This means that a 100 kΩ resistor (or smaller value, like 10 kΩ) needs to be connected between the SD signal and Ground.  Most breakout boards based on the INMP441 part will already have this resistor installed, so, most often you do not need to worry about this.  But, if you see random data in the lowest byte then it likely means this resistor needs to be added.

For most applications 24 bits of audio sample resolution is not needed.  16 bits is almost always good enough.  For example, "CD quality" is only 16 bits for each sample.  Unless you need 24 bits of resolution for a special signal processing application I recommend setting `bits=16` in the constructor.  This will improve the performance of the data transfer as 1/2 the amount sample data is copied into the receiving bytearray buffer.

**Reply by walklinewang** [reply]
@miketeachman thanks for your tips about sound quality, I can understand what @saraverbeecke means and I have the same doubt, whether `inmp441` or `ics-43434`, their datasheet has written the feather of "High-Precision 24-Bit Data", I have tried set `bits=16` to read buffered data and directly output them via `max98357`, sounds good, but is this the correct way to transfer data? In other words, does it cause loss any details of sound?
